### ISSUE TITLE
To many epochs

### ISSUE THREAD

**Issue by david-waterworth** [BatsResearch/wiser#30]
This may be specific to HMM, I've not checked LinkedHMM

`train_generative_model` loops on epochs

https://github.com/BatsResearch/wiser/blob/0ea47ec6211e2d779f34aebd847a5a11018cb6d7/wiser/generative/util.py#L72-L75

`HMM.estimate_label_model` calls `_do_estimate_label_model`

https://github.com/BatsResearch/labelmodels/blob/3b5683bef7ee902dc205a8cd93572b05b88dbad5/labelmodels/hmm.py#L110

`_do_estimate_label_model` contains a full epoch / batch training loop

https://github.com/BatsResearch/labelmodels/blob/3b5683bef7ee902dc205a8cd93572b05b88dbad5/labelmodels/label_model.py#L72-L92

End result is epoch^2 iterations, and it looks like the weights are reset so the end results is 5 epochs repeated 5 times rather than 25 epochs

> INFO:root:Random seed: 0
> INFO:root:Epoch 1/5
> INFO:root:Train Loss: -7093.234375
> INFO:root:Epoch 2/5
> INFO:root:Train Loss: -7093.293945
> INFO:root:Epoch 3/5
> INFO:root:Train Loss: -7093.284180
> INFO:root:Epoch 4/5
> INFO:root:Train Loss: -7075.308594
> INFO:root:Epoch 5/5
> INFO:root:Train Loss: -7074.095703
> INFO:root:Random seed: 0
> INFO:root:Epoch 1/5
> INFO:root:Train Loss: -7093.329102
> INFO:root:Epoch 2/5
> INFO:root:Train Loss: -7086.911621
> INFO:root:Epoch 3/5
> INFO:root:Train Loss: -7082.193359
> INFO:root:Epoch 4/5
> INFO:root:Train Loss: -7067.540039
> INFO:root:Epoch 5/5
> INFO:root:Train Loss: -7067.520508
> INFO:root:Random seed: 0
> INFO:root:Epoch 1/5
> INFO:root:Train Loss: -7093.300293
>

**Comment by ChaceHayhurst** [comment]
Hi, in this code segment we were trying to evaluate the model after every epoch, and use the best params we see (because the last epoch doesn't necessarily yield the best params). This is the reason for the loop in util.py. Unfortunately, we forgot to set config.epochs to 1 before going into this loop which produces the behavior you're seeing (fixed in pr https://github.com/BatsResearch/wiser/pull/32). I don't see anywhere in estimate_label_model or _do_estimate_label_model that resets the params between calls, so this should be working as intended now.  Thanks for pointing this out!
